Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,701 @@
+
+```qsharp
+open Microsoft.Quantum.Arrays;
+
+let perfect = [
+    [PauliI, PauliX, PauliZ, PauliZ, PauliX],
+    [PauliX, PauliZ, PauliZ, PauliX, PauliI],
+    [PauliZ, PauliZ, PauliX, PauliI, PauliX],
+    [PauliZ, PauliX, PauliI, PauliX, PauliZ]
+];
+// perfect: Pauli[][]
+let syndrome = ForEach(Measure(_, register), perfect);
+// syndrome: Result[]
+```
+
+Example 3:
+Representing the generators of a stabilizer group as a sparse array of Pauli operators and their indices.
+
+```qsharp
+let shor = [
+    [(PauliX, 0), (PauliX, 1)],
+    [(PauliX, 1), (PauliX, 2)],
+    [(PauliX, 3), (PauliX, 4)],
+    [(PauliX, 4), (PauliX, 5)],
+    [(PauliX, 6), (PauliX, 7)],
+    [(PauliX, 7), (PauliX, 8)],
+
+    [(PauliZ, 0), (PauliZ, 1), (PauliZ, 2), (PauliZ, 3), (PauliZ, 4), (PauliZ, 6)],
+    [(PauliZ, 3), (PauliZ, 4), (PauliZ, 6), (PauliZ, 7), (PauliZ, 8), (PauliZ, 9)],
+];
+// shor: (Pauli, Int)[][]
+```
+
+Example 4:
+A prior distribution over phases represented as an array of double-precision real numbers.
+
+```qsharp
+open Microsoft.Quantum.Arrays;
+open Microsoft.Quantum.Convert;
+
+// suppose likelihood is a function of type Int -> Double.
+let nPoints = 1000;
+let prior = ConstantArray(nPoints, 1.0 / IntAsDouble(nPoints));
+mutable posterior = prior;
+for ((idx, pr) in Enumerated(posterior)) {
+    set posterior w/= idx <- posterior[idx] * likelihood(idx);
+}
+set posterior = PNormalized(posterior, 1.0);
+```
+
+Example 5:
+A function that returns an elementary matrix (that is, a matrix with a one in the (𝑖, 𝑗)th element and zeros elsewhere).
+
+```qsharp
+open Microsoft.Quantum.Arrays;
+
+function ElementaryMatrix<'T>(
+    (nRows : Int, nColumns : Int),
+    (idxRow : Int, idxCol : Int),
+    zero : 'T, one : 'T
+) : 'T[][] {
+    mutable matrix = ConstantArray(
+        nRows,
+        ConstantArray(nCols, zero)
+    );
+    return matrix w/ row <- (matrix[row] w/ col <- one);
+}
+
+let e01 = ElementaryMatrix((2, 2), (0, 1), 0.0, 1.0);
+// e01: Double[][] = [[0.0, 1.0], [0.0, 0.0]]
+```
+
+## Proposed Modification
+
+Building on the utility of 1-D array notation, this suggestion proposes modifying Q# to include new multidimensional array types `[|'T`|], `[||'T||]`, and so forth. Like values of type `'T[]`, these new multidimensional would also be immutable, and could be manipulated by using the subscript (`[]`) and copy-and-update (`w/`) operators.
+
+For any type `'T`, this proposal introduces a new two-dimensional array type `[|'T|]`, a three-dimensional array type `[||'T||]`, and so forth.
+For consistency with these new types, this proposal also introduces `['T]` as alternative notation for `'T[]`; we will use new notation for `'T[]` in the remainder of the proposal.
+
+New values of type `[|'T|]` can be written as literals using `[ [| ... |] ]` delimiters instead of `[]`, while literals of type `[||'T||]` can be written using `[ [| [| ... |] |] ]` delimiters (see example 1 below).
```

Comment:
It would be good to add a sentence to this paragraph that they can also be constructed using library functions, listing the public functions for array construction further below with the list of internal library functions.